src/new/wallet.js

```diff
@@ -2469,6 +2464,285 @@ class HathorWallet extends EventEmitter {
+      for (const output of tx.outputs) {
+        if (output.value > MAX_OUTPUT_VALUE) {
+          throw new OutputValueError(`Transaction with id ${tx.tx_id} has output value of ${helpers.prettyValue(output.value)}. Maximum value is ${helpers.prettyValue(MAX_OUTPUT_VALUE)}`);
+        }
+      }
+
+      // We have an output field 'spent_by' that is filled everytime we receive a tx from the websocket that spends the output
+      // Between the tx creation and the websocket message being received we might select an utxo that had already been selected before
+      // To prevent this from happening we've created a custom field ('selected_as_input') that we set when we select the utxo to be used in a tx
+      // The if...else below if to keep this custom attribute correct even after receiving new data from an old tx
+      if (tx.tx_id in historyTransactions) {
+        // It's not a new tx
+        const storageTx = historyTransactions[tx.tx_id];
+        for (const [index, output] of tx.outputs.entries()) {
+          output['selected_as_input'] = storageTx.outputs[index]['selected_as_input'];
+        }
+      } else {
+        // It's a new tx
+        for (const output of tx.outputs) {
+          output['selected_as_input'] = false;
+        }
+      }
+
+      historyTransactions[tx.tx_id] = tx
+
+      for (const txin of tx.inputs) {
+        const key = walletData.keys[txin.decoded.address];
+        if (key) {
+          allTokens.add(txin.token);
+          if (key.index > maxIndex) {
+            maxIndex = key.index;
+            lastUsedAddress = txin.decoded.address
+          }
+        }
+      }
+      for (const txout of tx.outputs) {
+        const key = walletData.keys[txout.decoded.address];
+        if (key) {
+          allTokens.add(txout.token);
+          if (key.index > maxIndex) {
+            maxIndex = key.index;
+            lastUsedAddress = txout.decoded.address
+          }
+        }
+      }
+    }
+
+    let lastUsedIndex = wallet.getLastUsedIndex();
+    if (lastUsedIndex === null) {
+      lastUsedIndex = -1;
+    }
+
+    let lastSharedIndex = wallet.getLastSharedIndex();
+    if (lastSharedIndex === null) {
+      lastSharedIndex = -1;
+    }
+
+    if (maxIndex > lastUsedIndex && lastUsedAddress !== null) {
+      // Setting last used index and last shared index
+      wallet.setLastUsedIndex(lastUsedAddress);
+      // Setting last shared address, if necessary
+      const candidateIndex = maxIndex + 1;
+      if (candidateIndex > lastSharedIndex) {
+        const address = this.getAddressAtIndex(candidateIndex);
+        wallet.updateAddress(address, candidateIndex);
+      }
+    }
+
+    // Saving to storage before resolving the promise
+    wallet.saveAddressHistory(historyTransactions, allTokens);
+
+    const lastGeneratedIndex = wallet.getLastGeneratedIndex();
+
+    return { addressesFound: lastGeneratedIndex + 1, maxIndex, historyLength: Object.keys(historyTransactions).length };
```

Comment:
`newSharedAddress` and `newSharedIndex` that were also returned in the old utils are not used?